File: src/ch1/caesar_enc.py
# Caesar Cipher Implementation
# ord(a) => 97 // Ordinal 즉 순서로써 a 의 유니코드 "순서" 정도로 생각하면 된다.
# chr(97) => a
#
a = 'HELLO'
def cypher(text, shift):
    result = ""
    for char in text:
        if char.isalpha():
            shift_base = ord('A') if char.isupper() else ord('a')
            result += chr((ord(char) - shift_base + shift) % 26 + shift_base)
        else:
            result += char
    return result

def encrypt(text, shift):
    # 'A'와 'Z'의 문자 코드를 취득
    code_a = ord('A')
    code_z = ord('Z')
    # 결과를 대입할 변수
    result = ""
    for ch in text:
        # 문자 코드로 변환
        code = ord(ch)
        # A-Z 사이인가?
        if code_a <= code <= code_z:
          code = (code - code_a + shift) % 26 + code_a # %26 을 쓰는 이유는 XYZ => ABC 로 보여야 하기에.
        # 문자로 변환
        result += chr(code)
    return result


def encrypt_graceful(text, shift):
  a = ord('A')
  conv = lambda n: chr((ord(n) - a + shift) % 26 + a)
  enc1 = lambda n: conv(n) if 'A' <= n <= 'Z' else n
  return ''.join([enc1(n) for n in text])


if __name__ == "__main__":
  enc = encrypt("I LOVE RUST.", 3)
  dec = encrypt(enc, -3)
  print(enc, "=>", dec)

  enc2 = encrypt_graceful("I LOVE RUST.", 3)
  dec2 = encrypt_graceful(enc2, -3)
  print(enc2, "=>", dec2)

File: src/ch1/caesar_enc.rs
// 암호화 함수
fn encrypt_copilot(text: &str, shift: i16) -> String {
    let mut encrypted_text = String::new();
    for c in text.chars() {
        if c.is_ascii_alphabetic() {
            let base = if c.is_ascii_uppercase() { 'A' } else { 'a' } as u8;
            let shifted = ((c as u8 - base + shift as u8) % 26 + base) as char;
            encrypted_text.push(shifted);
        } else {
            encrypted_text.push(c);
        }
    }
    encrypted_text
}

fn encrypt(text: &str, shift: i16) -> String {
  // 'A'와 'Z'의 문자 코드를 i16 타입으로 취득
  let code_a = 'A' as i16;
  let code_z = 'Z' as i16;
  // 암호화된 문자열을 저장할 변수
  let mut result = String::new();
  for ch in text.chars() {
    let mut code = ch as i16;
    // A부터 Z까지의 문자라면
    if code_a <= code && code <= code_z {
      code = (code - code_a + shift + 26) % 26 + code_a;
    }      // 문자 코드를 문자로 변환하여 결과에 추가
    // 문자 코드를 다시 문자로 변환
    result.push((code as u8) as char);
  }
  return result;
}

fn multiply(a: i16, b: i16) -> i16 {
  a * b // 끝이 ;으로 끝나지 않는군. js 의 => 함수란 비슷하네.
}

fn main() {
  let enc = encrypt("I LOVE RUST", 3);
  let dec = encrypt(&enc, -3);
  println!("{} => {}", enc, dec);

  // let ex1 = multiply(3, 5);
  // println!("3 * 5 = {}", ex1);
  // let enc2 = encrypt_copilot("I LOVE RUST", 3);
  // let dec2 = encrypt_copilot(&enc2, -3); // 문자열을 String &str
  // println!("{} => {}", enc2, dec2);

  // 클로저
  let x2 = |n| n*2;
  println!("{}", x2(2));
  println!("{}", x2(8)); // => 16
}